Polio 670 destruction update

quang-le opened the pull request:
There was only 1 destruction of vaccines per rounds, but we need many

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] Did I add translations
- [X] My migrations file are included

## Changes

- Created a `Destruction` model
- Updated `RoundSerializer` to use the new model
- Created `DestructionForm` and `DestructionsForm` to handle adding and removing destructions
- Updated yup validation to reflect changes
- fixed some translation typos
- removed validator error on number fields with value 0
- Add fix for `TextArea` infinite loop bug


## How to test

Go to campaigns, open one, go Vaccines tab and try adding and removing destructions

## Print screen / video

https://user-images.githubusercontent.com/38907762/200021794-18db2189-fd96-4a3d-a847-603ab2d2f880.mov


## Notes

There's room for improvement as most code (both front and back) is pretty much copy pasted from Shipments

quang-le committed:
POLIO-670: comment out code

quang-le committed:
Merge branch 'POLIO-669_shipments_update' of https://github.com/BLSQ/iaso into POLIO-670_destruction_update

quang-le committed:
POLIO-670: add Destruction model

quang-le committed:
POLIO-670: update translations

quang-le committed:
POLIO-670: format migration

quang-le committed:
POLIO-670: update form fro multiple destructions

quang-le committed:
POLIO-670: add validation for destruction

quang-le committed:
POLIO-670: don't error number fields on 0

quang-le committed:
POLIO-670 add missing migration

hakifran commented on plugins/polio/js/src/constants/translations/en.json:
```diff
@@ -1,4 +1,5 @@
 {
+    "Date of the destruction report": "Date of the destruction report",
```
This shouldn't be there.

hakifran commented on plugins/polio/js/src/constants/translations/fr.json:
```diff
@@ -103,6 +103,7 @@
     "iaso.polio.import_line_list": "Importer Line List",
     "iaso.polio.import_line_list.successMessage": "Import réussi pour {amount} campagne(s)",
     "iaso.polio.label.add": "Ajouter",
+    "iaso.polio.label.addDestruction": "Ajouter destruction",
```
ce ne serait pas mieux si on disait "Ajouter une destruction"?

hakifran commented on plugins/polio/js/src/constants/translations/fr.json:
```diff
@@ -286,6 +289,7 @@
     "iaso.polio.label.rejectBudgetForCampaign": "Rejeter le budget pour la campagne: {campaign}",
     "iaso.polio.label.rejected": "Rejeté",
     "iaso.polio.label.Religious_leader": "Leader religieux",
+    "iaso.polio.label.removeLastDestruction": "Effacer dernière destruction",
```
"Effacer la dernière destruction"

hakifran commented on plugins/polio/js/src/forms/DestructionsForm.tsx:
```diff
@@ -0,0 +1,108 @@
+
+    const lastIndex = destructions.length >= 1 && destructions.length - 1;
+
+    const handleAddShipment = () => {
```
This method should be renamed to "handleAddDestruction"

quang-le replied:
Me copy/pasting my own code?😅

hakifran commented on plugins/polio/js/src/forms/DestructionsForm.tsx:
```diff
@@ -0,0 +1,108 @@
+        setFieldValue(`${accessor}.destructions`, newShipments);
+    };
+
+    const handleRemoveLastShipment = useCallback(() => {
```
This method too "handleRemoveLastDestruction"

hakifran commented on plugins/polio/js/src/forms/DestructionsForm.tsx:
```diff
@@ -0,0 +1,108 @@
+    };
+
+    const handleRemoveLastShipment = useCallback(() => {
+        const newShipments = [...destructions];
```
Here too newDestructions

hakifran commented on plugins/polio/js/src/forms/VaccineManagementForm.tsx:
```diff
@@ -192,11 +193,16 @@ export const VaccineManangementForm: FunctionComponent<Props> = () => {
+                    accessor={accessor}
+                    round={rounds[roundIndex]}
+                />
+                {/* <Grid container direction="row" spacing={2} item xs={12}>
```
This commented code be removed to make the code more clean. thank you

hakifran reviewed (commented):
Apart from some translation and methods name to change, it seems working for me and the remaining code seems to me correct. thank you @quang-le

quang-le committed:
POLIO-670: code review @hakifran

quang-le committed:
POLIO-670: fix TextArea debounce bug

- set default bounce time to 0
- wrap callbacks in useCallback in Budget and Storage

beygorghor reviewed (changes requested):
<img width="1345" alt="Screenshot 2022-11-07 at 12 07 32" src="https://user-images.githubusercontent.com/12494624/200295790-e3fe6553-356c-4835-a2b7-a32552f556f3.png">
<img width="1323" alt="Screenshot 2022-11-07 at 12 07 09" src="https://user-images.githubusercontent.com/12494624/200295876-58665d76-aee3-4d77-b194-b01894beb9a9.png">

I have an error on textArea while i shouldn't
<img width="1094" alt="Screenshot 2022-11-07 at 12 07 22" src="https://user-images.githubusercontent.com/12494624/200295913-d8c18cc4-d6ce-48f2-bb7e-6094d7ff27d4.png">
Also adding some more space before the divider if possible

quang-le committed:
POLIO-670: review @beygorghor

quang-le commented:
> 

It seems validation for Shipments broke somewhere on the way

codeKameleon commented:
@quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving

codeKameleon commented on plugins/polio/js/src/forms/DestructionsForm.tsx:
```diff
@@ -0,0 +1,108 @@
+        setFieldValue(`${accessor}.destructions`, newDestructions);
+    };
+
+    const handleRemoveLastDestruction = useCallback(() => {
```
@quang-le Is there a use case where the user would like to delete a older destruction ? For the moment the feature only allow to remove the last one

quang-le replied:
Not really. They might want to edit it, but once the vaccines are destroyed, they are destroyed. Same with Shipments

quang-le commented:
> @quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving

The thing is that it's conditional: all destruction inputs should be filled for a line (except comment) only if one has been filled. This as per client's request. Same with Shipments

quang-le commented:
> > 
> 
> It seems validation for Shipments broke somewhere on the way

It's not that, it's just that for new campaigns, the whole form will error when there's no obr_name. It works fine otherwise

codeKameleon commented:
> > @quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving
> 
> The thing is that it's conditional: all destruction inputs should be filled for a line (except comment) only if one has been filled. This as per client's request. Same with Shipments

Ok i see, so maybe in a future PR, show an info message in the Destruction section if the user put a value on one of the inputs, we could show "These fiellds need to be filled ..."

quang-le committed:
Merge branch 'POLIO-667-vaccine-formA-changes' of https://github.com/BLSQ/iaso into POLIO-670_destruction_update

olethanh commented:
EDIT : Apparently I cannot read. You already did the AuditDestructionSerializer, you just need to add it to the AuditRoundSerializer

You need to add the new Models to the Audit so the change modifications get kept.

See following PR for example on how to do this
https://github.com/BLSQ/iaso/pull/60/files

But basically you need to ad an AuditSerializer and then reference it in the

quang-le committed:
POLIO-670: fix textarea state bug

quang-le committed:
POLIO-670 add AuditDestructionSerializer to AuditRound

quang-le committed:
POLIO-670 fix serializer typo

hakifran committed:
put style propriety in same place

hakifran reviewed (approved):
Seems good to me. thank you @quang-le

beygorghor committed:
use a map

beygorghor committed:
remove console. log
